### PR TITLE
support uint16/32, and tests added

### PR DESCRIPTION
**Pull request by Lavishq**
https://github.com/karalabe/ssz/issues/12

image
image

- [x] Added uint16 support
- [x] Added uint32 support
- [x] tests

### CODE REVIEW

**Comment by karalabe** [comment]
Sweet, very nice :)

One more thing to make it perfect if you're up for it: tests :D But don't worry, it's easy because the test data is there, it just needs to be called:

- This one was added for the uint8 test https://github.com/karalabe/ssz/blob/main/tests/testtypes/consensus-spec-tests/types_basics.go#L12
- There are test types for uint16 and uint32 too as seen in https://github.com/ethereum/consensus-specs/blob/e9be4e2e3aeb8d0fbe3bff800328445e21a1fdf3/tests/generators/ssz_generic/ssz_container.py#L14
- Please get the `SmallTestStruct` and `FixedTestStruct` from that second python link into Go as per the first link.
- Add the code generator calls like https://github.com/karalabe/ssz/blob/main/tests/testtypes/consensus-spec-tests/types_basics.go#L9
- Run `go generate` to create the type defs in the test folder
- Add the two new types to the tests in https://github.com/karalabe/ssz/blob/main/tests/consensus_specs_test.go#L50

If you're feeling a bit more adventurous, you can also add `VarTestStruct` and `ComplexTestStruct`, which isn't much harder, just the fields types are not only primitives any more, rather you'll need to dig a bit deeper. But if you don't want to, I'm happy with the first too types only too :)

**Comment by Lavishq** [comment]
is this code okay for tests? im asking bc im writing go for the first time and have only done tour of go upto variables many month back
https://github.com/karalabe/ssz/commit/d8b729a287a2b07ebf29b6ae0cae2472df7ffcd0

also tried to add `VarTestStruct` and `ComplexTestStruct` tests

```go
type VarTestStruct struct {
	A [1]byte `ssz-size:"2" ssz:"bits"`
	// B [1]byte `ssz-size:"8" ssz:"bits"`
	C [1]byte `ssz-size:"1" ssz:"bits"`
	// A: uint16
	// B: List[uint16, 1024]
	// C: uint8
}

type ComplexTestStruct struct {
	A [1]byte `ssz-size:"2" ssz:"bits"`
	// B: List[uint16, 128]
	C [1]byte `ssz-size:"1" ssz:"bits"`
	// D: ByteList[256]
	// E: VarTestStruct
	// F: Vector[FixedTestStruct, 4]
	// G: Vector[VarTestStruct, 2]
}
```
but since my knowledge on dynamic variables in golang is limited `i failed` but i can try `VarTestStruct` and `ComplexTestStruct` tomoro

if there is something incorrect do lmk, since every line of code was done by the resources provided and some pattern matching across codebase and couldnt find anything to match from gobyexample

i can make the pr of `SmallTestStruct` and `FixedTestStruct` rn

File: tests/testtypes/consensus-spec-tests/types_basics.go [hunk 1]
```diff
@@ -7,12 +7,25 @@ package consensus_spec_tests
...
+type SmallTestStruct struct {
+	A [1]byte `ssz-size:"2" ssz:"bits"`
+	B [1]byte `ssz-size:"2" ssz:"bits"`
+}
+
+type FixedTestStruct struct {
+	A [1]byte `ssz-size:"1" ssz:"bits"`
+	B [1]byte `ssz-size:"8" ssz:"bits"`
+	C [1]byte `ssz-size:"4" ssz:"bits"`
+}
```
Comment (karalabe) [inline tests/testtypes/consensus-spec-tests/types_basics.go:27]: Ah, no no 

The types of the fields are

```
class SmallTestStruct(Container):
    A: uint16
    B: uint16


class FixedTestStruct(Container):
    A: uint8
    B: uint64
    C: uint32
```

I.e. 

```go
type SmallTestStruct struct {
	A uint16
	B uint16
}

type FixedTestStruct struct {
	A uint8
	B uint64
	C uint32
}
```

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:27]: You don't need any struct tags on there. Those are more for the weird types :D

Comment (Lavishq) [reply tests/testtypes/consensus-spec-tests/types_basics.go:27]: i tried these but they dont generate tests when i ran `go generate` with errors 

`/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_fixed_test_struct_ssz.go:15:2: T (type [8]byte) does not satisfy ssz.commonBitsLengths`
`/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_fixed_test_struct_ssz.go:16:2: T (type [4]byte) does not satisfy ssz.commonBitsLengths`

so i tinkered to remove the errors^
would fix asap

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:45]: Ah, damn, didn't see that there's a list of uint16 in there in VarTestStruct. Yeah, just drop that type and the ComplexTestStruct. List of primitives needs a lot more work there so it's very out of scope for this PR. Just leave SmallTestStruct and FixedTestStruct and shoult be fine like that.

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:45]: Also pls add the test call for the two types here https://github.com/karalabe/ssz/blob/main/tests/consensus_specs_test.go#L51 

That should make this PR complete :)

File: tests/consensus_specs_test.go [hunk 2]
```diff
@@ -49,6 +49,8 @@ func commonPrefix(a []byte, b []byte) []byte {
 // consensus spec tests repo and runs the encoding/decoding/hashing round.
 func TestConsensusSpecBasics(t *testing.T) {
 	testConsensusSpecBasicType[*types.SingleFieldTestStruct](t, "SingleFieldTestStruct")
+	testConsensusSpecBasicType[*types.SingleFieldTestStruct](t, "SmallTestStruct")
+	testConsensusSpecBasicType[*types.SingleFieldTestStruct](t, "FixedTestStruct")
```
Comment (karalabe) [inline tests/consensus_specs_test.go:53]: Heh, you need to change the type name too :P `types.SingleFieldTestStruct` -> your two new types :)

Comment (Lavishq) [reply tests/consensus_specs_test.go]: really, sorry for this 
thank you for not getting annoyed on so so silly issues and guiding in a detailed manner

Comment (karalabe) [reply tests/consensus_specs_test.go]: It's ok :) You're sacrificing your time on this :)

File: codec.go [hunk 3]
```diff
@@ -79,6 +79,32 @@ func DefineUint8[T ~uint8](c *Codec, n *T) {
...
+		return
+	}
+	if c.dec != nil {
+		DecodeUint16(c.dec, n)
+		return
+	}
+	HashUint16(c.has, *n)
+}
+
+// DefineUint16 defines the next field as a uint16.
```
Comment (karalabe) [inline codec.go:95]: Pls change to 32 both values

File: cmd/sszgen/opset.go [hunk 4]
```diff
@@ -73,6 +73,26 @@ func (p *parseContext) resolveBasicOpset(typ *types.Basic, tags *sizeTag) (opset
 			"DecodeUint8({{.Codec}}, &{{.Field}})",
 			[]int{1},
 		}, nil
+	case types.Uint16:
+		if tags != nil && tags.size[0] != 2 {
+			return nil, fmt.Errorf("byte basic type requires ssz-size=2: have %d", tags.size[0])
```
Comment (karalabe) [inline cmd/sszgen/opset.go:78]: Pls replace byte in the error message to uint16

File: cmd/sszgen/opset.go [hunk 5]
```diff
@@ -73,6 +73,26 @@ func (p *parseContext) resolveBasicOpset(typ *types.Basic, tags *sizeTag) (opset
...
+		}
+		return &opsetStatic{
+			"DefineUint16({{.Codec}}, &{{.Field}})",
+			"EncodeUint16({{.Codec}}, &{{.Field}})",
+			"DecodeUint16({{.Codec}}, &{{.Field}})",
+			[]int{2},
+		}, nil
+	case types.Uint32:
+		if tags != nil && tags.size[0] != 4 {
+			return nil, fmt.Errorf("byte basic type requires ssz-size=4: have %d", tags.size[0])
```
Comment (karalabe) [inline cmd/sszgen/opset.go:88]: Pls replace byte in the error message to uint32

File: tests/testtypes/consensus-spec-tests/types_basics.go [hunk 6]
```diff
@@ -7,12 +7,25 @@ package consensus_spec_tests
 import "github.com/prysmaticlabs/go-bitfield"
 
 //go:generate go run -cover ../../../cmd/sszgen -type SingleFieldTestStruct -out gen_single_field_test_struct_ssz.go
+//go:generate go run -cover ../../../cmd/sszgen -type SmallTestStruct -out gen_small_test_struct_ssz.go
+//go:generate go run -cover ../../../cmd/sszgen -type FixedTestStruct -out gen_fixed_test_struct_ssz.go
```
Comment (karalabe) [inline tests/testtypes/consensus-spec-tests/types_basics.go:11]: I think you forgot to regenerate the files after chaning the fields to the uintxx types. Pls do that too.

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: Just run `go generate` and should do the trick

Comment (Lavishq) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: ❯ go generate
warning: GOCOVERDIR not set, no coverage data emitted
2024/07/26 21:55:23 Generating SSZ bindings for: [SingleFieldTestStruct]
/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_fixed_test_struct_ssz.go:14:2: T (type uint8) does not satisfy ssz.commonBitsLengths
/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_fixed_test_struct_ssz.go:15:2: T (type uint64) does not satisfy ssz.commonBitsLengths
/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_fixed_test_struct_ssz.go:16:2: T (type uint32) does not satisfy ssz.commonBitsLengths
/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_small_test_struct_ssz.go:14:2: T (type uint16) does not satisfy ssz.commonBitsLengths
/home/lavishq/ssz/tests/testtypes/consensus-spec-tests/gen_small_test_struct_ssz.go:15:2: T (type uint16) does not satisfy ssz.commonBitsLengths
package github.com/karalabe/ssz/tests/testtypes/consensus-spec-tests has errors
exit status 1
types_basics.go:9: running "go": exit status 1

this error

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: Yeah, you'll have to delete the two files. The generator tries to load the package to get the type and there's a build error so it aborts.

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: You probably generate the fields originally with a bad field, and then you left the two files there, so changing the fields always resulted in an error from the leftover old generated code.

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: Ideally the generator would "ignore" that one file it's generating a replacement for, but there's no easy way to do that, so this is something I have to live with.

Comment (Lavishq) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: after deleting those files and running `go generate` 
tests are proceeding :)

so i will commit and push that once

Comment (karalabe) [reply tests/testtypes/consensus-spec-tests/types_basics.go:11]: Very nice :). Those aren't the tests mind you, just the types used by the tests, but nonetheless it's great :)

**Review by karalabe — APPROVED** [review]
All tests green. Thank you for the PR. LGTM! :)

**Comment by karalabe** [comment]
@Lavishq Just one more tip in general, in Go, you can run all the tests from all packages via `go test ./...`. The one you did running `go test` only ran it for tests in the root folder, which aren't really many/any. This is generally not the case for Go packages of course, but for this specific project there is such an extensive external test suite, that it made little sense to redo internal unit tests when I could just rely on the external suite.

Thanks again :)

**Comment by Lavishq** [comment]
hm, i know nothing about go, i once did tour of go upto variables and imports, probably a little more for about 20-30 mins  and that is all [ also dont remember much ]
so i do need to learn it the right way, recently was looking at op-geth and didnt understand much, thought that since it is gfg i would quickly learn from gobyexample or the codebase itself ^^

thanks ser, for bearing with me, it was embarrasing to make silly mistakes but was a learning exp and would do again but maybe after i finish tour of go this time :)

**Pull request merged**